### PR TITLE
refactor: split client and server errors

### LINKED ISSUE #1088: split `jsonrpsee::core::Error` to `ClientError` and `ServerError` error types
Currently jsonrpsee has a big error type which has many variants that are not applicable for the server and vice verse for the client.

Thus, we should try to split it up to help for make it easier to understand and to help dealing with https://github.com/paritytech/jsonrpsee/issues/1067

### PR DESCRIPTION
**Pull request by niklasad1**
Close #1088

### CODE REVIEW

File: server/src/future.rs [hunk 1]
```diff
@@ -61,6 +60,11 @@ impl StopHandle {
 	}
 }
 
+#[derive(Debug, Copy, Clone, thiserror::Error)]
+#[error("The server is already stopped")]
+pub struct AlreadyStoppedError;
+
+
```
Comment (niklasad1) [inline server/src/future.rs:67]:
```suggestion
```

File: client/http-client/src/client.rs [hunk 2]
```diff
@@ -327,7 +327,8 @@ where
 	where
 		R: DeserializeOwned + fmt::Debug + 'a,
 	{
-		let batch = batch.build()?;
+		// TODO fix unwrap.
+		let batch = batch.build().unwrap();
```
Comment (niklasad1) [inline client/http-client/src/client.rs:331]:
```suggestion
		let batch = batch.build()?;
```

File: client/http-client/src/client.rs [hunk 3]
```diff
@@ -327,6 +327,7 @@ where
 	where
 		R: DeserializeOwned + fmt::Debug + 'a,
 	{
+		// TODO fix unwrap.
```
Comment (niklasad1) [inline client/http-client/src/client.rs:330]:
```suggestion
```

File: core/src/client/async_client/helpers.rs [hunk 4]
```diff
@@ -198,7 +197,7 @@ pub(crate) fn process_single_response(
 			let sub_id = match sub_id {
 				Ok(Some(sub_id)) => sub_id,
 				Ok(None) => {
-					let _ = send_back_oneshot.send(Err(Error::InvalidSubscriptionId));
+					let _ = send_back_oneshot.send(Err(Error::InvalidRequestId(InvalidRequestId::Invalid("Invalid subscription ID".into()))));
```
Comment (niklasad1) [inline core/src/client/async_client/helpers.rs:200]:
```suggestion
					let _ = send_back_oneshot.send(Err(Error::InvalidSubscriptionId));
```

File: core/src/client/async_client/mod.rs [hunk 5]
```diff
@@ -427,6 +427,7 @@ impl ClientT for Client {
 	where
 		R: DeserializeOwned,
 	{
+		// TODO: remove unwrap
```
Comment (niklasad1) [inline core/src/client/async_client/mod.rs:430]:
```suggestion
```

File: client/ws-client/src/lib.rs [hunk 6]
```diff
@@ -216,11 +216,7 @@ impl WsClientBuilder {
 	}
 
 	/// Build the [`WsClient`] with specified [`TransportSenderT`] [`TransportReceiverT`] parameters
-	///
-	/// ## Panics
```
Comment (lexnv) [inline client/ws-client/src/lib.rs:220]: nit: Wouldn't this panic? (build_with_stream and build are having the panic section)

**Review by lexnv — APPROVED** [review]
LGTM!

File: core/src/lib.rs [hunk 7]
```diff
@@ -54,12 +54,13 @@ cfg_server! {
...
 cfg_client! {
 	pub mod client;
+	pub use client::Error as ClientError;
 }
 
 /// Shared tracing helpers to trace RPC calls.
 pub mod tracing;
 pub use async_trait::async_trait;
-pub use error::{Error, GenericTransportError, StringError};
+//pub use error::{Error, GenericTransportError, StringError};
```
Comment (niklasad1) [inline core/src/lib.rs:63]:
```suggestion
pub use error::{GenericTransportError, StringError, RegisterMethodError};
```

**Review by jsdw — APPROVED** [review]

**Pull request merged**